pgx-pg-sys/build.rs

```diff
@@ -392,23 +416,632 @@ fn impl_pg_node(
+    {
+        quote! {
+            impl pg_sys::PgNode for List {
+                fn traverse<T>(&mut self, walker_fn: fn(*mut Node, &mut T) -> (), context: &mut T) {
+                    if self.length == 0 { return; }
+                    let mut cell = unsafe { *self.head };
+                    for _index in 0..self.length {
+                        let item = unsafe { cell.data.ptr_value as *mut Node };
+                        if !item.is_null() {
+                            walker_fn(item, context);
+                            Node::traverse::<T>( unsafe { &mut *(item as *mut Node) }, walker_fn, context);
+                        }
+                        if !cell.next.is_null() {
+                            cell = unsafe { *cell.next };
+                        }
+                    }
+                }
+            }
+            impl std::fmt::Display for List {
+                fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                    write!(f, "[")?;
+                    if self.length != 0 {
+                        let mut cell = unsafe { *self.head };
+                        for index in 0..self.length {
+                            if index != 0 {
+                                write!(f, ", ")?;
+                            }
+                            let item = unsafe { cell.data.ptr_value as *mut Node };
+                            if item.is_null() {
+                                write!(f, "(null)")?;
+                            } else {
+                                write!(f, "{}", unsafe { *item })?;
+                            }
+                            if !cell.next.is_null() {
+                                cell = unsafe { *cell.next };
+                            }
+                        }
+                    }
+                    write!(f, "]")
+                }
+            }
+        }
+    } else {
+        quote! {
+            impl pg_sys::PgNode for List {
+                fn traverse<T>(&mut self, walker_fn: fn(*mut Node, &mut T) -> (), context: &mut T) {
+                    for index in 0..self.length {
+                        let item = unsafe { (*self.elements.offset(index as isize)).ptr_value as *mut Node };
+                        if !item.is_null() {
+                            walker_fn(item, context);
+                            Node::traverse::<T>( unsafe { &mut *(item as *mut Node) }, walker_fn, context);
+                        }
+                    }
+                }
+            }
+            impl std::fmt::Display for List {
+                fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                    write!(f, "[")?;
+                    for index in 0..self.length {
+                        if index != 0 {
+                            write!(f, ", ")?;
+                        }
+                        let item = unsafe { (*self.elements.offset(index as isize)).ptr_value as *mut Node };
```

Comment:
It looks like this code assumes that all of the entire length of `self.elements` is valid to access, and then does so, even if every single element it encounters is itself a null pointer. It should be stating that precondition once, using [`slice::from_raw_parts`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html) and then iterating that directly.